src/main/clojure/clara/tools/loop_detector.cljc

```diff
@@ -0,0 +1,102 @@
+(ns clara.tools.loop-detector
+  (:require [clara.rules.listener :as l]
+            [clara.rules.engine :as eng]
+            [clara.tools.tracing :as trace]))
+
+(deftype CyclicalRuleListener [cycles-count max-cycles on-limit-fn]
+    l/ITransientEventListener
+  (left-activate! [listener node tokens]
+    listener)
+  (left-retract! [listener node tokens]
+    listener)
+  (right-activate! [listener node elements]
+    listener)
+  (right-retract! [listener node elements]
+    listener)
+  (insert-facts! [listener node token facts]
+    listener)
+  (alpha-activate! [listener node facts]
+    listener)
+  (insert-facts-logical! [listener node token facts]
+    listener)
+  (retract-facts! [listener node token facts]
+    listener)
+  (alpha-retract! [listener node facts]
+    listener)
+  (retract-facts-logical! [listener node token facts]
+    listener)
+  (add-accum-reduced! [listener node join-bindings result fact-bindings]
+    listener)
+  (remove-accum-reduced! [listener node join-bindings fact-bindings]
+    listener)
+  (add-activations! [listener node activations]
+    listener)
+  (remove-activations! [listener node activations]
+    listener)
+  (fire-activation! [listener activation resulting-operations]
+    listener)
+  (fire-rules! [listener node]
+    listener)
+  (activation-group-transition! [listener original-group new-group]
+    (when (>= @cycles-count max-cycles)
+      (on-limit-fn))
+    (swap! cycles-count inc))
+  (to-persistent! [listener]
+    (CyclicalRuleListener. (atom 0) max-cycles on-limit-fn))
+
+  l/IPersistentEventListener
+  (to-transient [listener]
+    listener))
+
+(defn throw-exception-on-max-cycles
+  []
+  (let [trace (trace/listener->trace (l/to-persistent! (:listener eng/*current-session*)))]
+    (throw (ex-info "Reached maximum activation group transitions threshhold; an infinite loop is suspected"
+                    (cond-> {:clara-rules/infinite-loop-suspected true}
+                      trace (assoc :trace trace))))))
+
+(defn ->standard-out-warning
+  []
+  (println "Reached maximum activation group transitions threshhold; an infinite loop is suspected"))
+
+(defn ->invoke-once-fn
+  [wrapped-fn]
+  (let [warned (atom false)]
+    (fn []
+      (when-not @warned
+        (reset! warned true)
+        (wrapped-fn)))))
```

Comment:
This is new from the last PR.  This allows us to restrict user-provided limit functions to a single invocation, unlike the previous one where they would be invoked many times once the limit is exceeded.

Reply:
I replaced this with a delay in https://github.com/cerner/clara-rules/pull/458/commits/26d967ebbd659a669207334b84281a925663ec64

I also noticed that there was a bug where the atom with the cycles count could end up being shared between sessions.  I fixed this and added a test for the scenario.